main/init/buildgen/src/mill/main/buildgen/BuildGenBase.scala

```diff
@@ -4,26 +4,29 @@ import mill.main.buildgen.BuildGenUtil.{buildPackages, compactBuildTree, writeBu
 
 import scala.collection.immutable.SortedMap
 
-trait BuildGenBase[M, D] {
+trait BuildGenBase[M, D, I] {
```

Comment:
What does this new type parameter do? I don't see us using it anywhere

Reply:
In my current information sbt's `IrBaseInfo` is not retrieved from a project but from `mill.main.sbt.BuildInfo`, which is extracted from sbt's `ThisBuild`, so for this to work the input type `I` has to be `(BuildInfo, Tree[Node[Project]])` instead of just `Tree[Node[Project]]`.

Reply:
What do you think of converting all these type parameters into abstract type members? See the TODO in the code for more:

https://github.com/com-lihaoyi/mill/blob/22033e909dd461689b374815ca4602b087b8830f/main/init/buildgen/src/mill/main/buildgen/BuildGenBase.scala#L7-L23

Reply:
I think leaving them as type parameters is fine for now. In the end it doesn't really matter, and although such a conversion may or may not be desirable, it's unrelated to SBT and can come in a follow up